### PR TITLE
Make the public ui.Codec API Future based instead of callback based.

### PR DESCRIPTION
**Pull request by amirh**

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -896,25 +896,35 @@ abstract class Codec extends NativeFieldWrapperClass2 {
...
-  String getNextFrame(NextFrameInfoCallback callback) native "Codec_getNextFrame";
+  String _getNextFrame(NextFrameInfoCallback callback) native "Codec_getNextFrame";
 
   /// Release the resources used by this object. The object is no longer usable
   /// after this method is called.
   void dispose() native "Codec_dispose";
 }
 
-/// Callback signature for [imageCodecFromList].
+/// Callback signature for [_instantiateImageCodec].
```
Comment (Hixie) [inline lib/ui/painting.dart]: If it's a signature only for a private class, the typedef itself should also be private.

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -896,25 +896,35 @@ abstract class Codec extends NativeFieldWrapperClass2 {
   /// Fetches the next animation frame.
   ///
   /// Wraps back to the first frame after returning the last frame.
-  ///
+  Future<FrameInfo> getNextFrame() async {
```
Comment (Hixie) [inline lib/ui/painting.dart]: docs should probably mention that the future can complete with an error.

Comment (Hixie) [inline lib/ui/painting.dart]: this function shouldn't be `async`, since it doesn't use `await`.

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -896,25 +896,35 @@ abstract class Codec extends NativeFieldWrapperClass2 {
...
+/// Callback signature for [_instantiateImageCodec].
 ///
-/// If parsing the data passed to [imageCodecFromList] failed, the result will
+/// If parsing the data passed to [_instantiateImageCodec] failed, the result will
 /// be null.
 typedef void CodecCallback(Codec result);
 
-/// Instatiates a [Codec] object for an image binary data.
+/// Instantiates a [Codec] object for an image binary data.
+Future<Codec> instantiateImageCodec(Uint8List list) async {
```
Comment (Hixie) [inline lib/ui/painting.dart]: this function shouldn't be `async`, since it doesn't use `await`.

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -896,25 +896,35 @@ abstract class Codec extends NativeFieldWrapperClass2 {
...
-/// Callback signature for [imageCodecFromList].
+/// Callback signature for [_instantiateImageCodec].
 ///
-/// If parsing the data passed to [imageCodecFromList] failed, the result will
+/// If parsing the data passed to [_instantiateImageCodec] failed, the result will
 /// be null.
 typedef void CodecCallback(Codec result);
 
-/// Instatiates a [Codec] object for an image binary data.
+/// Instantiates a [Codec] object for an image binary data.
```
Comment (Hixie) [inline lib/ui/painting.dart]: these docs should be more detailed (e.g. what is "an image binary data"?).

**Comment by Hixie** [comment]
LGTM, but travis is angry.

File: lib/ui/painting.dart (hunk 1 above)
Comment (amirh) [reply lib/ui/painting.dart]: changed to use  _Callback<Codec> instead

**Pull request merged**